File: Python/algorithms/sorting/heap_sort.py
"""
Heap Sort Algorithm
Time Complexity: O(n log n) in all cases
Space Complexity: O(1) - sorts in place
Author: Karanjot Singh
Date: October 2025
Hacktoberfest 2025
"""

def heapify(arr, n, i):
    """
    Heapify subtree rooted at index i

    Args:
        arr: Array to heapify
        n: Size of heap
        i: Root index of subtree
    """
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    # Check if left child exists and is greater than root
    if left < n and arr[left] > arr[largest]:
        largest = left

    # Check if right child exists and is greater than largest so far
    if right < n and arr[right] > arr[largest]:
        largest = right

    # If largest is not root, swap and continue heapifying
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


def heap_sort(arr):
    """
    Heap sort implementation

    Steps:
    1. Build max heap from array
    2. Extract elements one by one from heap
    3. Place extracted element at the end

    Args:
        arr: List to sort (modified in place)
    """
    n = len(arr)

    # Build max heap
    # Start from last non-leaf node and heapify each node
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Extract elements from heap one by one
    for i in range(n - 1, 0, -1):
        # Move current root to end
        arr[0], arr[i] = arr[i], arr[0]

        # Heapify the reduced heap
        heapify(arr, i, 0)


def heap_sort_copy(arr):
    """
    Heap sort that returns sorted copy

    Args:
        arr: List to sort

    Returns:
        Sorted copy of the list
    """
    arr_copy = arr.copy()
    heap_sort(arr_copy)
    return arr_copy


# Test cases
if __name__ == "__main__":
    print("=== Heap Sort Algorithm ===\n")

    # Test case 1: Random array
    test1 = [12, 11, 13, 5, 6, 7]
    print(f"Original array: {test1}")
    sorted1 = heap_sort_copy(test1)
    print(f"Sorted array:   {sorted1}")

    # Test case 2: Already sorted
    test2 = [1, 2, 3, 4, 5]
    print(f"\nOriginal array: {test2}")
    sorted2 = heap_sort_copy(test2)
    print(f"Sorted array:   {sorted2}")

    # Test case 3: Reverse sorted
    test3 = [10, 8, 6, 4, 2]
    print(f"\nOriginal array: {test3}")
    sorted3 = heap_sort_copy(test3)
    print(f"Sorted array:   {sorted3}")

    # Test case 4: Array with duplicates
    test4 = [4, 10, 3, 5, 1, 3, 10]
    print(f"\nOriginal array: {test4}")
    sorted4 = heap_sort_copy(test4)
    print(f"Sorted array:   {sorted4}")

    # Test case 5: Large array
    test5 = [64, 34, 25, 12, 22, 11, 90, 88, 45, 50, 23, 36]
    print(f"\nOriginal array: {test5}")
    sorted5 = heap_sort_copy(test5)
    print(f"Sorted array:   {sorted5}")

    # Test case 6: Single element
    test6 = [42]
    print(f"\nOriginal array: {test6}")
    sorted6 = heap_sort_copy(test6)
    print(f"Sorted array:   {sorted6}")

    print("\n✅ All test cases completed!")

File: C/algorithms/sorting/heap_sort.c
/*
 * Algorithm: Heap Sort
 * Description: In-place comparison-based sorting algorithm using a binary heap (max-heap).
 * Time Complexity: O(n log n) for all cases
 * Space Complexity: O(1) auxiliary (in-place)
 * Author: rishanmenezes
 * Date: 2025-10-05
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

/*
 * Sifts an element down the heap to maintain max-heap property.
 * @param arr: Array representing the heap
 * @param n: Current heap size
 * @param i: Index to heapify at
 */
static void heapify(int arr[], int n, int i) {
    // Input validation: indexes within bounds
    if (arr == NULL || n <= 0 || i < 0 || i >= n) {
        return; // Nothing to do for invalid input
    }

    int largest = i;            // Assume current node is largest
    int left = 2 * i + 1;       // Left child
    int right = 2 * i + 2;      // Right child

    // If left child is larger than root
    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    // If right child is larger than largest so far
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    // If largest is not root
    if (largest != i) {
        int tmp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = tmp;

        // Recursively heapify the affected sub-tree
        heapify(arr, n, largest);
    }
}

/**
 * Builds a max-heap from an unsorted array in O(n) time.
 */
static void build_max_heap(int arr[], int n) {
    if (arr == NULL || n <= 1) return;

    // Start from last internal node and heapify downwards
    for (int i = (n / 2) - 1; i >= 0; --i) {
        heapify(arr, n, i);
    }
}

/**
 * Sorts the array in ascending order using heap sort.
 * @param arr: Array to sort (modified in place)
 * @param n: Number of elements in arr
 * @return: 0 on success, -1 for invalid input
 */
int heap_sort(int arr[], int n) {
    // Input validation
    if (arr == NULL || n < 0) {
        fprintf(stderr, "Error: Invalid input to heap_sort()\n");
        return -1;
    }
    if (n <= 1) {
        return 0; // Already sorted or nothing to do
    }

    // Step 1: Build max-heap
    build_max_heap(arr, n);

    // Step 2: Repeatedly extract the maximum to the end
    for (int end = n - 1; end > 0; --end) {
        // Move current root (max) to end
        int tmp = arr[0];
        arr[0] = arr[end];
        arr[end] = tmp;

        // Heapify reduced heap
        heapify(arr, end, 0);
    }

    return 0;
}

/* -------------------------- Test Utilities --------------------------- */

static void print_array(const char* label, const int arr[], int n) {
    printf("%s[", label);
    for (int i = 0; i < n; ++i) {
        printf("%d%s", arr[i], (i + 1 == n) ? "" : ", ");
    }
    printf("]\n");
}

static bool is_sorted_non_decreasing(const int arr[], int n) {
    for (int i = 1; i < n; ++i) {
        if (arr[i - 1] > arr[i]) return false;
    }
    return true;
}

static void run_test_case(const char* name, int input[], int n) {
    printf("\n== %s ==\n", name);
    print_array("Input: ", input, n);
    int rc = heap_sort(input, n);
    if (rc != 0) {
        printf("heap_sort returned error code %d\n", rc);
        return;
    }
    print_array("Sorted: ", input, n);
    printf("Correct: %s\n", is_sorted_non_decreasing(input, n) ? "true" : "false");
}

int main(void) {
    // Test 1: Typical unsorted array
    int a1[] = {64, 34, 25, 12, 22, 11, 90};
    run_test_case("Typical", a1, (int)(sizeof(a1) / sizeof(a1[0])));

    // Test 2: Already sorted array
    int a2[] = {1, 2, 3, 4, 5, 6};
    run_test_case("Already Sorted", a2, (int)(sizeof(a2) / sizeof(a2[0])));

    // Test 3: Reverse sorted array
    int a3[] = {9, 8, 7, 6, 5, 4, 3, 2, 1};
    run_test_case("Reverse Sorted", a3, (int)(sizeof(a3) / sizeof(a3[0])));

    // Test 4: Duplicates and negatives
    int a4[] = {5, -1, 3, 5, 2, 2, -3, 0};
    run_test_case("Duplicates & Negatives", a4, (int)(sizeof(a4) / sizeof(a4[0])));

    // Test 5: Single element
    int a5[] = {42};
    run_test_case("Single Element", a5, (int)(sizeof(a5) / sizeof(a5[0])));

    // Test 6: Empty array (n = 0) - should be a no-op
    int a6[] = {};
    run_test_case("Empty Array", a6, 0);

    return 0;
}
